android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt

```diff
@@ -66,54 +79,31 @@ class RemoteSettingsClient(
      */
     suspend fun write(response: RemoteSettingsResponse): RemoteSettingsResult = withContext(Dispatchers.IO) {
         try {
-            val jsonObject = JSONObject()
-            jsonObject.put("response", response)
-            AtomicFile(file).writeString { jsonObject.toString() }
+            val jsonString = Json.encodeToString(
+                SerializableRemoteSettingsResponse.serializer(),
+                response.toSerializable(),
+            )
+            AtomicFile(file).writeString { jsonString }
             RemoteSettingsResult.Success(response)
         } catch (e: IOException) {
             RemoteSettingsResult.DiskFailure(e)
         }
     }
 
-    /**
-     * Updates the local storage with [response] data if [response.lastModified] is more recent than [lastModified]
-     *
-     * @param response is collections data from remote server or local storage.
-     * @param lastModified tells when the last change was made to the response.
-     */
-    suspend fun writeIfModifiedSince(
-        response: RemoteSettingsResponse,
-        lastModified: ULong,
-    ): RemoteSettingsResult =
-        withContext(Dispatchers.IO) {
-            try {
-                if (response.lastModified > lastModified) {
-                    write(response)
-                } else {
-                    RemoteSettingsResult.Success(response)
-                }
-            } catch (e: IOException) {
-                RemoteSettingsResult.DiskFailure(e)
-            }
-        }
-
     /**
      * Reads all response for a collection found in the local storage.
      */
     suspend fun read(): RemoteSettingsResult = withContext(Dispatchers.IO) {
-        if (!file.exists()) {
-            RemoteSettingsResult.DiskFailure(FileNotFoundException("File not found"))
-        } else {
-            try {
-                val fileReader = FileReader(file)
-                val jsonObject = JSONObject(fileReader.readText())
-                val response = jsonObject.get("response") as RemoteSettingsResponse
-
-                fileReader.close()
-                RemoteSettingsResult.Success(response)
-            } catch (e: IOException) {
-                RemoteSettingsResult.DiskFailure(e)
+        try {
+            if (!file.exists()) {
+                RemoteSettingsResult.DiskFailure(FileNotFoundException("File not found"))
+            } else {
+                val jsonString = file.readText()
+                val response = Json.decodeFromString<SerializableRemoteSettingsResponse>(jsonString)
+                RemoteSettingsResult.Success(response.toRemoteSettingsResponse())
             }
+        } catch (e: IOException) {
+            RemoteSettingsResult.DiskFailure(e)
```

Comment:
We could add another `catch` here for a [serialization exception](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-serialization-exception/) if we want to avoid crashes for malformed data. Should be pretty low risk since we're also the ones writing to disk though.